src/web/mjs/engine/DiscordPresence.mjs

```diff
@@ -0,0 +1,167 @@
+    _onMangaClicked(e) {
+        // Hentai check
+        if(e.detail.connector.tags.includes('hentai') || e.detail.connector.tags.includes('porn')) {
+            this.hentai = true;
+        } else {
+            this.hentai = false;
+        }
+
+        this.status['details'] = 'Browsing ' + e.detail.connector.label;
+        this.status['state'] = 'Looking at ' + e.detail.title;
+        this.status.startTimestamp = + new Date();
+        this.statusNew = true;
+        if (this.enabled && !this.rpc) this.startDiscordPresence();
+    }
+
+    _onChapterGetPages(e) {
+        // Hentai check
+        if(e.detail.manga.connector.tags.includes('hentai') || e.detail.manga.connector.tags.includes('porn')) {
+            this.hentai = true;
+        } else {
+            this.hentai = false;
+        }
+
+        this.status['details'] = 'Reading ' + e.detail.manga.title;
+        this.status['state'] = e.detail.title;
+        this.status.startTimestamp = + new Date();
+        this.statusNew = true;
+        if (this.enabled && !this.rpc) this.startDiscordPresence();
+    }
+
+    async updateStatus() {
+        if(this.rpc) {
+            if (this.enabled && this.statusNew) {
+                this.IpcBytes = this.rpc.transport.socket.bytesWritten;
+                if( !this.hentai || (this.hentai && this.enabledHentai)) {
+                    this.rpc.setActivity(this.status);
+                } else {
+                    this.statusNew = false;
+                }
+            }
+
+            // Test if IPC is still active
+            if ( this.statusNew && this.rpc.transport.socket.bytesWritten > this.IpcBytes) {
+                this.IpcBytes = this.rpc.transport.socket.bytesWritten;
+                this.statusNew = false;
+            } else if (this.rpc.transport.socket.bytesWritten == this.IpcBytes && this.statusNew) {
+                console.warn('WARNING: DiscordPresence - Lost connection to Discord.');
+                this.stopDiscordPresence();
+            }
+        }
+    }
+
+    stopDiscordPresence() {
+        this.statusNew = false;
+        if (this.rpc) {
+            this.rpc.clearActivity();
+            this.rpc.destroy();
+            this.rpc = null;
+        }
+    }
+
+    async startDiscordPresence() {
+        if (!this.rpc) {
+            this.rpc = new DiscordRPC.Client({ transport: 'ipc' });
+            this.rpc.on('ready', () => {
+                this.status.startTimestamp = + new Date();
+                
+                // some delay for Discord to be receptive
+                setTimeout( () => {
+                    this.updateStatus();
+                }, 2000);
+
+                // activity can only be set every 15 seconds
+                setInterval(() => {
+                    this.updateStatus();
+                }, 15200);
+            });
+            this.rpc.login({ clientId: discordPresenceId })
+            .catch (error => {
+                if (error.message == 'Could not connect') {
```

Comment:
Is this check reliable? Is the text always english and case sensitive?
Maybe check `error.code` if available such as 4001, 4002 and whatever the discord codes are...
https://discord.com/developers/docs/topics/opcodes-and-status-codes#gateway-gateway-close-event-codes

Reply:
Unfortunately discord-rpc does not expose an error code:
![image](https://user-images.githubusercontent.com/18627623/86147221-6a398900-baf9-11ea-8a19-e919198f10ed.png)
Also this is an error that triggers if there is no IPC connection to Discord possible, so no Discord API can be leveraged. But i changed it to `error.message.match(/[A-z]+/g).join('').toUpperCase() == 'COULDNOTCONNECT'` to make a it a little bit more robust.

Reply:
Also a nice example for short Regex.test
`if(/could\s*not\s*connect/i.test(error.message)) { ... error }`